heat/core/linalg/qr.py

```diff
@@ -4,1039 +4,338 @@
+            # this corresponds to the loop over all columns in classical Gram-Schmidt
+            if i < nprocs - 1:
+                k_loc_i = min(A.shape[0], A.lshape_map[i, 1])
+                Q_buf = torch.zeros(
+                    (A.shape[0], k_loc_i), dtype=A.larray.dtype, device=A.device.torch_device
+                )
 
-    upp = r[: upper.shape[0]]
-    low = r[upper.shape[0] :]
-    if rank == pr0:
-        r_tiles.arr.lloc[upper_inds[0] : upper_inds[1], upper_inds[2] : upper_inds[3]] = upp
-    else:  # rank == pr1:
-        r_tiles.arr.lloc[lower_inds[0] : lower_inds[1], lower_inds[2] : lower_inds[3]] = low
+            if A.comm.rank == i:
+                # orthogonalize the current block of columns by utilizing PyTorch QR
+                Q_curr, R_loc = torch.linalg.qr(A_columns, mode="reduced")
+                if i < nprocs - 1:
+                    Q_buf = Q_curr
+                if mode == "reduced":
+                    Q.larray = Q_curr
+                r_size = R.larray[R_shapes[i] : R_shapes[i + 1], :].shape[0]
+                R.larray[R_shapes[i] : R_shapes[i + 1], :] = R_loc[:r_size, :]
+
+            if i < nprocs - 1:
+                # broadcast the orthogonalized block of columns to all other processes
+                req = A.comm.Ibcast(Q_buf, root=i)
+                req.Wait()
+
+            if A.comm.rank > i:
+                # subtract the contribution of the current block of columns from the remaining columns
+                R_loc = Q_buf.T @ A_columns
+                A_columns -= Q_buf @ R_loc
+                r_size = R.larray[R_shapes[i] : R_shapes[i + 1], :].shape[0]
+                R.larray[R_shapes[i] : R_shapes[i + 1], :] = R_loc[:r_size, :]
+
+        if mode == "reduced":
+            Q = Q[:, :k].balance()
+        else:
+            Q = None
 
-    if column < r_tiles.tile_columns - 1:
-        upper_rest_size = (upper_size[0], r_tiles.arr.gshape[1] - upper_inds[3])
-        lower_rest_size = (lower_size[0], r_tiles.arr.gshape[1] - lower_inds[3])
+        return QR(Q, R)
 
-        if rank == pr0:
-            upper_rest = r_tiles.arr.lloc[upper_inds[0] : upper_inds[1], upper_inds[3] :]
-            lower_rest = torch.zeros(
-                lower_rest_size, dtype=r_tiles.arr.dtype.torch_type(), device=a_torch_device
+    if A.split == 0:
+        # implementation of TS-QR for split = 0
+        # check that data distribution is reasonable for TS-QR (i.e. tall-skinny matrix with also tall-skinny local chunks of data)
+        if A.lshape_map[:, 0].max().item() < A.shape[1]:
+            raise ValueError(
+                "A is split along the rows and the local chunks of data are rectangular with more rows than columns. \n Applying TS-QR in this situation is not reasonable w.r.t. runtime and memory consumption. \n We recomment to split A along the columns instead. \n In case this is not an option for you, please open an issue on GitHub."
```

Comment:
This is going to be weird for the users. You can start with a global TS matrix and end up with non-TS local arrays if you use enough processes. We need to explain it better / introduce an earlier check, maybe make a helper function available that calculates the most reasonable split for  given matrix size and nr of processes? I'll think about it some more.